exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/ConvertMetadataAggregateToDirectScanRule.java

```diff
@@ -0,0 +1,278 @@
+        MetadataAggregateContext updatedContext = agg.getContext().toBuilder()
+            .createNewAggregations(false)
+            .build();
+        converted = new MetadataAggRel(agg.getCluster(), agg.getTraitSet(), converted, updatedContext);
+      }
+
+      call.transformTo(converted);
+    } catch (Exception e) {
+      logger.warn("Unable to use parquet metadata for ANALYZE due to exception {}", e.getMessage(), e);
+    }
+  }
+
+  private DirectGroupScan buildDirectScan(List<SchemaPath> interestingColumns, DrillScanRel scan, PlannerSettings settings) throws IOException {
+    DrillTable drillTable = Utilities.getDrillTable(scan.getTable());
+
+    Map<String, Class<?>> schema = new HashMap<>();
+
+    // populates schema and fieldIndexes to be used when adding record values
+    FormatSelection selection = (FormatSelection) drillTable.getSelection();
+    // adds partition columns to the schema
+    for (String partitionColumnName : ColumnExplorer.getPartitionColumnNames(selection.getSelection(), settings.getOptions())) {
+      schema.put(partitionColumnName, String.class);
+    }
+
+    String rgi = settings.getOptions().getOption(ExecConstants.IMPLICIT_ROW_GROUP_INDEX_COLUMN_LABEL).string_val;
+    String rgs = settings.getOptions().getOption(ExecConstants.IMPLICIT_ROW_GROUP_START_COLUMN_LABEL).string_val;
+    String rgl = settings.getOptions().getOption(ExecConstants.IMPLICIT_ROW_GROUP_LENGTH_COLUMN_LABEL).string_val;
+    String lmt = settings.getOptions().getOption(ExecConstants.IMPLICIT_LAST_MODIFIED_TIME_COLUMN_LABEL).string_val;
+
+    // adds internal implicit columns to the schema
+
+    schema.put(MetastoreAnalyzeConstants.SCHEMA_FIELD, String.class);
+    schema.put(MetastoreAnalyzeConstants.LOCATION_FIELD, String.class);
+    schema.put(rgi, String.class);
+    schema.put(rgs, String.class);
+    schema.put(rgl, String.class);
+    schema.put(lmt, String.class);
+
+    return populateRecords(interestingColumns, schema, scan, settings);
+  }
+
+  /**
+   * Populates records list with row group metadata.
+   */
+  private DirectGroupScan populateRecords(Collection<SchemaPath> interestingColumns, Map<String, Class<?>> schema,
+      DrillScanRel scan, PlannerSettings settings) throws IOException {
+    ParquetGroupScan parquetGroupScan = (ParquetGroupScan) scan.getGroupScan();
+    DrillTable drillTable = Utilities.getDrillTable(scan.getTable());
+
+    Multimap<Path, RowGroupMetadata> rowGroupsMetadataMap = parquetGroupScan.getMetadataProvider().getRowGroupsMetadataMap();
+
+    Table<String, Integer, Object> recordsTable = HashBasedTable.create();
+    FormatSelection selection = (FormatSelection) drillTable.getSelection();
+    List<String> partitionColumnNames = ColumnExplorer.getPartitionColumnNames(selection.getSelection(), settings.getOptions());
+
+    String rgi = settings.getOptions().getOption(ExecConstants.IMPLICIT_ROW_GROUP_INDEX_COLUMN_LABEL).string_val;
```

Comment:
Looks like are doing the same operation on getting special implicit columns names several times. It would be better to have holder where you'll pass `options`, which would do this implicitly. Similar as we have `org.apache.drill.exec.store.kafka.ReadOptions` in Kafka.